### ISSUE TITLE
Feature, methods diffInOpenHours and diffInOpenMinutes

### ISSUE THREAD

**Issue by FuriosoJack** [kylekatarnls/business-time#25]
Hello.
Feachure, is posible diffInOpenHours and diffInOpenMinutes?

**Comment by kylekatarnls** [comment]
It would be relevant to have such methods. I will not have the time right now to implement it. Maybe I could prepare unit tests. But it's open to pull-requests.

For consistency with existing methods in Carbon, we should have the following:

`floatDiffInOpenHours($date = null, bool $absolute = true, int $options = 0): float`
`diffInOpenHours($date = null, bool $absolute = true, int $options = 0): int`
`floatDiffInOpenMinutes($date = null, bool $absolute = true, int $options = 0): float`
`diffInOpenMinutes($date = null, bool $absolute = true, int $options = 0): int`

A few specifications:
- `$date` can be any value supported by `Carbon::diff*` methods (date/time string, DateTimeInterface, etc.), so it should use:
```php
$base = isset($this) ? $this : static::now();
/** @var CarbonInterface $date */
$date = $base->resolveCarbon($date);
```
- with `$absolute = false`, the result of `$a->floatDiffInOpenMinutes($b)` will be negative if `$b < $a`
- with `$absolute = true` (the default value), the result of `$a->floatDiffInOpenMinutes($b)` should be equal to `abs($a->floatDiffInOpenMinutes($b))`
To get a `CarbonInterface` from function parameter.
- `$options` is bits-flag that may contains `BusinessTime::HOLIDAYS_ARE_CLOSED` (the only option for now), this option make holidays automatically considered as closed even if not specified in hours definition.
- `$a->diffInOpenHours($b)` should be equivalent to `$a->floatDiffInOpenMinutes($b) / CarbonInterface::MINUTES_PER_HOUR`
- It would be fine to have smallest units seconds/milliseconds/microseconds (with a minimum of duplicated code in those methods => they can divide/multiple the result of a derived unit) but we should not implement bigger unit, as it may be confusing: either a day should be considered 1 per 24/hour, 1 per the current number of hours in the day/average hours in days, if it would use the `CarbonInterval::getCascadeFactors()` etc.
- On the top of that, we could have: `diffInOpenTime()` as a `CarbonInterval` and `diffInOpenTimeForHuman()` as a translated string such as `"2 hour and 30 minutes"`.
- `"Closed"` methods can be implemented and would be a priori Total-diff - Open-diff.
- The method should behave like `diffInReal*` methods of `Carbon`, e.g. always give the true elapsed time even where there is a DST change. But an option can be added to behave like `diffIn*` methods  (compensating DST changes).

**Comment by FuriosoJack** [comment]
at the moment i'm using

`php     
$diferenciaHorasCarbon = $fechaInicial->diffInHoursFiltered(function($date){          
  return $date->isOpen();
  },$fechaBaseFinal);`

and it seems to work. I base it on the diffInBusinessDays you have in the business-days package

**Comment by kylekatarnls** [comment]
This code is pretty fine if you don't have minutes in your open time ranges (e.g. calculate diff from 8:00 to 10:00 when the business open at 8:30), when having such ranges, this code will floor the result (or ceil for similar case in closing), or worst, it can floor N times if you have multiple ranges:
`'monday' => ['08:30-11:00', '12:30-17:00']`
your diff method applied to 8:00 to 18:00 will return 6 hours, while it's actually open 2.5 + 3.5 = 7 hours.

Second point is performance, this is iterating over every hours. `diffInBusinessDays` is basically not good at performance, but when iterating over hours, it becomes 24 times worst, if we do the same with minutes 1440 times worst, and here it can be very slow for long intervals (but as pointed above, we actually need minute precision to match at least the time range allowed precision).

`diffInBusinessDays` has the same imprecision level: it consider the dates you pass to it as "midnight" and ignore the time of the original dates. This is precise enough for business-days, since (as the name mention it), it has day-precision and in the scope of business-days, a day is either fully open or fully closed.

This precision will not be good enough for business-time, minute is the minimum, so we'll need to use nextOpen/nextClose, to skip loop turns. And ideally, we could be even more precise. Since we will not use a unit-based iteration, we will sum intervals and so it could be as precise as the given dates.

**Comment by kylekatarnls** [comment]
Here is the approach I have in mind (using pure-functions to simplify):

```php
function floatDiffInOpenMinutes(CarbonInterface $startDate, CarbonInterface $endDate): float {
  $time = 0;

  if ($endDate < $startDate) {
    return -floatDiffInOpenMinutes($endDate, $startDate);
  }

  $date = $startDate->copy();

  while ($date < $endDate) {
    if ($date->isClosed()) {
      $date = $date->nextOpen();
      continue;
    }

    $nextDate = $date->nextClose()->copy();
    $time += $date->floatDiffInRealSeconds(min($endDate, $nextDate));
    $date = $nextDate;
  }

  return $time / CarbonInterface::SECONDS_PER_MINUTE;
}

function floatDiffInOpenHours(CarbonInterface $startDate, CarbonInterface $endDate): float {
  return floatDiffInOpenMinutes($startDate, $endDate) / CarbonInterface::MINUTES_PER_HOUR;
}

function diffInOpenMinutes(CarbonInterface $startDate, CarbonInterface $endDate): int {
  return (int) floatDiffInOpenMinutes($startDate, $endDate);
}

function diffInOpenHours(CarbonInterface $startDate, CarbonInterface $endDate): int {
  return (int) floatDiffInOpenHours($startDate, $endDate);
}
```

**Comment by FuriosoJack** [comment]
Ok I am going to do some tests of the functions that you mentioned, with a project that I work which has specification of holidays according to my country (Colombia) and the opening hours of the business may be from just one day a week with a shortage of hours from 8am to 12pm or like every day of the week but with different business opening hours.

_Additionally, I would like to take the opportunity to ask you to believe that there is a problem in using `BusinessTime::enable ` with different configurations throughout the life of the request?._

**Comment by kylekatarnls** [comment]
Issues are also for bugs, you can create a new one, describe the steps I can follow to reproduce it with precise and static dates/input examples, show your input and what you expected to get. Thanks.

**Comment by kylekatarnls** [comment]
If you think you know how it should be fixed, then you can also do the same + a pull-request to propose your fix. That's open-source, as for this feature, anyone can contribute.

**Comment by kylekatarnls** [comment]
Correction:
```php
$time += $date->floatDiffInRealSeconds($nextDate);
```
must be:
```php
$time += $date->floatDiffInRealSeconds(min($endDate, $nextDate));
```

**Comment by kylekatarnls** [comment]
While working on this, I realized diff/floatDiff variants lived long enough. Truncate/cast the result is not the responsibility of the library. Carbon 3.0 will just provide the exact result of the calculation and let the user convert it on need.

business-time will anticipate this behavior and will simply provide "diff" methods that all return `float` results.

**Comment by FuriosoJack** [comment]
Hello, It is not reliable. The hours do not match.
Configuration


```php

[
   "monday" => [
      "08:00-18:00" 
   ], 
   "tuesday" => [
      "08:00-18:00" 
   ], 
   "wednesday" => [
      "08:00-18:00" 
   ], 
   "thursday" => [
      "08:00-18:00" 
   ], 
   "friday" => [
      "08:00-18:10" 
   ], 
   "saturday" => [
      "09:00-12:00" 
   ], 
   "sunday" => [
   ], 
   "holidaysAreClosed" => true, 
   "holidays" => [
      "region" => "es-national", 
      "with" => [
         "new-year" => null, 
         "substitutes-01-01" => null, 
         "01-06" => null, 
         "substitutes-01-06" => null, 
         "03-19" => null, 
         "substitutes-03-19" => null, 
         "easter" => null, 
         "substitutes-05-01" => null, 
         "1st-sunday-of-may" => null, 
         "easter-49" => null, 
         "07-25" => null, 
         "08-15" => null, 
         "substitutes-08-15" => null, 
         "10-12" => null, 
         "11-01" => null, 
         "substitutes-11-01" => null, 
         "12-06" => null, 
         "substitutes-12-08" => null, 
         "substitutes-12-25" => null, 
         "dia-ascension" => "= easter 43", 
         "corpus-christi" => "= easter 64", 
         "sagrado-corazon" => "= easter 71", 
         "dia-independencia" => "07-20", 
         "batalla-boyaca" => "08-07", 
         "año-nuevo" => "01-01", 
         "reyes-magos" => "= 01-06 if not Monday then next Monday", 
         "san-jose" => "= 03-19 if not Monday then next Monday", 
         "sanpedro-sanpablo" => "= 06-29 if not Monday then next Monday", 
         "asuncion-virgen" => "= 08-15 if not Monday then next Monday", 
         "dia-raza" => "= 10-12 if not Monday then next Monday", 
         "todos-santos" => "= 11-01 if not Monday then next Monday", 
         "independencia-cartajena" => "= 11-11 if not Monday then next Monday" 
      ] 
   ] 
]
```

**initial date**
`2020-06-12 18:04:49`

**end date**
`2020-06-16 08:23:41`

**Expected time difference in float**
3.61667

**Difference of hours obtained in float**
9.46228


I think floatDiffInRealSeconds is not taking into account that Monday June 15 was a holiday. What is in the holidays specification of the configuration and what would be the `corpus-christi` with syntactic sugar` = easter 64`

**Comment by kylekatarnls** [comment]
Indeed it works only with `CarbonImmutable` for `Carbon` class, `$nextDate = $date->nextClose();` also need to be replaced with `$nextDate = $date->copy()->nextClose();`

**Comment by kylekatarnls** [comment]
Methods implemented in 1.7.0:
https://github.com/kylekatarnls/business-time#diffasbusinessinterval
